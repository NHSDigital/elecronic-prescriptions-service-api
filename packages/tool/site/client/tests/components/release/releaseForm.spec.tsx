import {render} from "@testing-library/react"
import {screen, waitFor} from "@testing-library/dom"
import pretty from "pretty"
import * as React from "react"
import {expect} from "@jest/globals"
import ReleaseForm from "../../../src/components/release/releaseForm"
import userEvent from "@testing-library/user-event"
import {MemoryRouter} from "react-router"

const prescriptionId = "7A9089-A83008-56A03J"

test("Fields default to current values for nominated release", async () => {
  const {container} = render(
    <MemoryRouter>
      <ReleaseForm onSubmit={jest.fn}/>
    </MemoryRouter>
  )

  const releaseTypeContainer = await screen.findByLabelText<HTMLElement>("Choose how you want to release prescription(s)")
  const releaseTypeRadios = releaseTypeContainer.getElementsByClassName("nhsuk-radios__input") as HTMLCollectionOf<HTMLInputElement>
  expect(releaseTypeRadios).toHaveLength(3)
  expect(releaseTypeRadios[0].value).toEqual("all")
  expect(releaseTypeRadios[0].checked).toBeTruthy()
  expect(releaseTypeRadios[1].value).toEqual("prescriptionId")
  expect(releaseTypeRadios[1].checked).toBeFalsy()
  expect(releaseTypeRadios[2].value).toEqual("custom")
  expect(releaseTypeRadios[2].checked).toBeFalsy()

  const pharmacyContainer = await screen.findByLabelText<HTMLElement>("Pharmacy to release prescriptions to")
  const pharmacyRadios = pharmacyContainer.getElementsByClassName("nhsuk-radios__input") as HTMLCollectionOf<HTMLInputElement>
  expect(pharmacyRadios).toHaveLength(3)
  expect(pharmacyRadios[0].value).toEqual("VNFKT")
  expect(pharmacyRadios[0].checked).toBeFalsy()
  expect(pharmacyRadios[1].value).toEqual("YGM1E")
  expect(pharmacyRadios[1].checked).toBeFalsy()
  expect(pharmacyRadios[2].value).toEqual("custom")
  expect(pharmacyRadios[2].checked).toBeFalsy()

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Fields default to current values for patient release", async () => {
  const {container} = render(
    <MemoryRouter>
      <ReleaseForm prescriptionId={prescriptionId} onSubmit={jest.fn}/>
    </MemoryRouter>
  )

  const releaseTypeContainer = await screen.findByLabelText<HTMLElement>("Choose how you want to release prescription(s)")
  const releaseTypeRadios = releaseTypeContainer.getElementsByClassName("nhsuk-radios__input") as HTMLCollectionOf<HTMLInputElement>
  expect(releaseTypeRadios).toHaveLength(3)
  expect(releaseTypeRadios[0].value).toEqual("all")
  expect(releaseTypeRadios[0].checked).toBeFalsy()
  expect(releaseTypeRadios[1].value).toEqual("prescriptionId")
  expect(releaseTypeRadios[1].checked).toBeTruthy()
  expect(releaseTypeRadios[2].value).toEqual("custom")
  expect(releaseTypeRadios[2].checked).toBeFalsy()

  const pharmacyContainer = await screen.findByLabelText<HTMLElement>("Pharmacy to release prescriptions to")
  const pharmacyRadios = pharmacyContainer.getElementsByClassName("nhsuk-radios__input") as HTMLCollectionOf<HTMLInputElement>
  expect(pharmacyRadios).toHaveLength(3)
  expect(pharmacyRadios[0].value).toEqual("VNFKT")
  expect(pharmacyRadios[0].checked).toBeFalsy()
  expect(pharmacyRadios[1].value).toEqual("YGM1E")
  expect(pharmacyRadios[1].checked).toBeFalsy()
  expect(pharmacyRadios[2].value).toEqual("custom")
  expect(pharmacyRadios[2].checked).toBeFalsy()

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Prescription Id Field is required when releasing by prescription Id", async () => {
  const {container} = render(
    <MemoryRouter>
      <ReleaseForm onSubmit={jest.fn}/>
    </MemoryRouter>
  )

  const releaseTypeContainer = await screen.findByLabelText<HTMLElement>("Choose how you want to release prescription(s)")
  const releaseTypeRadios = releaseTypeContainer.getElementsByClassName("nhsuk-radios__input") as HTMLCollectionOf<HTMLInputElement>
  userEvent.click(releaseTypeRadios[1])
  userEvent.click(screen.getByText("Release"))

  await waitFor(() => screen.getByText("You must enter a 'Prescription ID' to release to when releasing a single prescription"))
  expect(screen.getByLabelText("Prescription ID")).toBeTruthy()

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Fhir Field is required when releasing a custom fhir release message", async () => {
  const {container} = render(
    <MemoryRouter>
      <ReleaseForm onSubmit={jest.fn}/>
    </MemoryRouter>
  )

  const releaseTypeContainer = await screen.findByLabelText<HTMLElement>("Choose how you want to release prescription(s)")
  const releaseTypeRadios = releaseTypeContainer.getElementsByClassName("nhsuk-radios__input") as HTMLCollectionOf<HTMLInputElement>
  userEvent.click(releaseTypeRadios[2])
  userEvent.click(screen.getByText("Release"))

  await waitFor(() => screen.getByText("You must enter a FHIR release message when selecting 'With a FHIR release message'"))
  expect(screen.getByLabelText("Paste a FHIR release message")).toBeTruthy()

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Pharmacy to release to is always required", async () => {
  const {container} = render(
    <MemoryRouter>
      <ReleaseForm onSubmit={jest.fn}/>
    </MemoryRouter>
  )

  userEvent.click(screen.getByText("Release"))

  await waitFor(() => screen.getByText("You must select a pharmacy to release to"))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
