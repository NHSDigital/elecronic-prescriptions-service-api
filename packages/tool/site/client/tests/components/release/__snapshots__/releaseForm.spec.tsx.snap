// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fhir Field is required when releasing a custom fhir release message 1`] = `
"<form>
  <div class="nhsuk-form-group nhsuk-form-group--error">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group nhsuk-form-group--error"><label class="nhsuk-label nhsuk-label--s" id="releaseType--label" for="releaseType">Choose how you want to release prescription(s)</label><span class="nhsuk-error-message" id="releaseType--error-message" role="alert"><span class="nhsuk-u-visually-hidden">Error: </span>You must enter a FHIR release message when selecting 'With a FHIR release message'</span>
        <div class="nhsuk-radios" id="releaseType" aria-labelledby="releaseType--label" value="custom">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-1" name="releaseType" aria-labelledby="releaseType-1--label" type="radio" value="all" checked=""><label class="nhsuk-label nhsuk-radios__label" id="releaseType-1--label" for="releaseType-1">All nominated prescriptions for the below pharmacy</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-2" name="releaseType" aria-labelledby="releaseType-2--label" type="radio" value="prescriptionId"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-2--label" for="releaseType-2">A single prescription by ID</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-3" name="releaseType" aria-labelledby="releaseType-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-3--label" for="releaseType-3">With a FHIR release message</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group"><label class="nhsuk-label" id="customReleaseFhir--label" for="customReleaseFhir">Paste a FHIR release message</label><textarea class="nhsuk-textarea" aria-labelledby="customReleaseFhir--label" name="customReleaseFhir" id="customReleaseFhir" rows="20"></textarea></div>
    </fieldset>
  </div>
  <div class="sc-bcPKBy hhHLxP"><button class="nhsuk-button" aria-disabled="false" type="submit">Release</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Fields default to current values for nominated release 1`] = `
"<form>
  <div class="nhsuk-form-group">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="releaseType--label" for="releaseType">Choose how you want to release prescription(s)</label>
        <div class="nhsuk-radios" id="releaseType" aria-labelledby="releaseType--label" value="all">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-1" name="releaseType" aria-labelledby="releaseType-1--label" type="radio" value="all" checked=""><label class="nhsuk-label nhsuk-radios__label" id="releaseType-1--label" for="releaseType-1">All nominated prescriptions for the below pharmacy</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-2" name="releaseType" aria-labelledby="releaseType-2--label" type="radio" value="prescriptionId"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-2--label" for="releaseType-2">A single prescription by ID</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-3" name="releaseType" aria-labelledby="releaseType-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-3--label" for="releaseType-3">With a FHIR release message</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="pharmacy--label" for="pharmacy">Pharmacy to release prescriptions to</label>
        <div class="nhsuk-radios" id="pharmacy" aria-labelledby="pharmacy--label" value="">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-1" name="pharmacy" aria-labelledby="pharmacy-1--label" type="radio" value="VNFKT"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-1--label" for="pharmacy-1">VNFKT - FIVE STAR HOMECARE LEEDS LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-2" name="pharmacy" aria-labelledby="pharmacy-2--label" type="radio" value="YGM1E"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-2--label" for="pharmacy-2">YGM1E - MBBM HEALTHCARE TECHNOLOGIES LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-3" name="pharmacy" aria-labelledby="pharmacy-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-3--label" for="pharmacy-3">Other</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class="sc-bcPKBy hhHLxP"><button class="nhsuk-button" aria-disabled="false" type="submit">Release</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Fields default to current values for patient release 1`] = `
"<form>
  <div class="nhsuk-form-group">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="releaseType--label" for="releaseType">Choose how you want to release prescription(s)</label>
        <div class="nhsuk-radios" id="releaseType" aria-labelledby="releaseType--label" value="prescriptionId">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-1" name="releaseType" aria-labelledby="releaseType-1--label" type="radio" value="all"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-1--label" for="releaseType-1">All nominated prescriptions for the below pharmacy</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-2" name="releaseType" aria-labelledby="releaseType-2--label" type="radio" value="prescriptionId" checked=""><label class="nhsuk-label nhsuk-radios__label" id="releaseType-2--label" for="releaseType-2">A single prescription by ID</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-3" name="releaseType" aria-labelledby="releaseType-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-3--label" for="releaseType-3">With a FHIR release message</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group"><label class="nhsuk-label" id="prescriptionId--label" for="prescriptionId">Prescription ID</label><input class="nhsuk-input nhsuk-input--width-30" aria-labelledby="prescriptionId--label" name="prescriptionId" id="prescriptionId" type="text" value="7A9089-A83008-56A03J"></div>
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="pharmacy--label" for="pharmacy">Pharmacy to release prescriptions to</label>
        <div class="nhsuk-radios" id="pharmacy" aria-labelledby="pharmacy--label" value="">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-1" name="pharmacy" aria-labelledby="pharmacy-1--label" type="radio" value="VNFKT"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-1--label" for="pharmacy-1">VNFKT - FIVE STAR HOMECARE LEEDS LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-2" name="pharmacy" aria-labelledby="pharmacy-2--label" type="radio" value="YGM1E"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-2--label" for="pharmacy-2">YGM1E - MBBM HEALTHCARE TECHNOLOGIES LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-3" name="pharmacy" aria-labelledby="pharmacy-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-3--label" for="pharmacy-3">Other</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class="sc-bcPKBy hhHLxP"><button class="nhsuk-button" aria-disabled="false" type="submit">Release</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Pharmacy to release to is always required 1`] = `
"<form>
  <div class="nhsuk-form-group nhsuk-form-group--error">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="releaseType--label" for="releaseType">Choose how you want to release prescription(s)</label>
        <div class="nhsuk-radios" id="releaseType" aria-labelledby="releaseType--label" value="all">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-1" name="releaseType" aria-labelledby="releaseType-1--label" type="radio" value="all" checked=""><label class="nhsuk-label nhsuk-radios__label" id="releaseType-1--label" for="releaseType-1">All nominated prescriptions for the below pharmacy</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-2" name="releaseType" aria-labelledby="releaseType-2--label" type="radio" value="prescriptionId"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-2--label" for="releaseType-2">A single prescription by ID</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-3" name="releaseType" aria-labelledby="releaseType-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-3--label" for="releaseType-3">With a FHIR release message</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group nhsuk-form-group--error"><label class="nhsuk-label nhsuk-label--s" id="pharmacy--label" for="pharmacy">Pharmacy to release prescriptions to</label><span class="nhsuk-error-message" id="pharmacy--error-message" role="alert"><span class="nhsuk-u-visually-hidden">Error: </span>You must select a pharmacy to release to</span>
        <div class="nhsuk-radios" id="pharmacy" aria-labelledby="pharmacy--label" value="">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-1" name="pharmacy" aria-labelledby="pharmacy-1--label" type="radio" value="VNFKT"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-1--label" for="pharmacy-1">VNFKT - FIVE STAR HOMECARE LEEDS LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-2" name="pharmacy" aria-labelledby="pharmacy-2--label" type="radio" value="YGM1E"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-2--label" for="pharmacy-2">YGM1E - MBBM HEALTHCARE TECHNOLOGIES LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-3" name="pharmacy" aria-labelledby="pharmacy-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-3--label" for="pharmacy-3">Other</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class="sc-bcPKBy hhHLxP"><button class="nhsuk-button" aria-disabled="false" type="submit">Release</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Prescription Id Field is required when releasing by prescription Id 1`] = `
"<form>
  <div class="nhsuk-form-group nhsuk-form-group--error">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group nhsuk-form-group--error"><label class="nhsuk-label nhsuk-label--s" id="releaseType--label" for="releaseType">Choose how you want to release prescription(s)</label><span class="nhsuk-error-message" id="releaseType--error-message" role="alert"><span class="nhsuk-u-visually-hidden">Error: </span>You must enter a 'Prescription ID' to release to when releasing a single prescription</span>
        <div class="nhsuk-radios" id="releaseType" aria-labelledby="releaseType--label" value="prescriptionId">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-1" name="releaseType" aria-labelledby="releaseType-1--label" type="radio" value="all" checked=""><label class="nhsuk-label nhsuk-radios__label" id="releaseType-1--label" for="releaseType-1">All nominated prescriptions for the below pharmacy</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-2" name="releaseType" aria-labelledby="releaseType-2--label" type="radio" value="prescriptionId"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-2--label" for="releaseType-2">A single prescription by ID</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="releaseType-3" name="releaseType" aria-labelledby="releaseType-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="releaseType-3--label" for="releaseType-3">With a FHIR release message</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group"><label class="nhsuk-label" id="prescriptionId--label" for="prescriptionId">Prescription ID</label><input class="nhsuk-input nhsuk-input--width-30" aria-labelledby="prescriptionId--label" name="prescriptionId" id="prescriptionId" type="text" value=""></div>
      <div class="nhsuk-form-group nhsuk-form-group--error"><label class="nhsuk-label nhsuk-label--s" id="pharmacy--label" for="pharmacy">Pharmacy to release prescriptions to</label><span class="nhsuk-error-message" id="pharmacy--error-message" role="alert"><span class="nhsuk-u-visually-hidden">Error: </span>You must select a pharmacy to release to</span>
        <div class="nhsuk-radios" id="pharmacy" aria-labelledby="pharmacy--label" value="">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-1" name="pharmacy" aria-labelledby="pharmacy-1--label" type="radio" value="VNFKT"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-1--label" for="pharmacy-1">VNFKT - FIVE STAR HOMECARE LEEDS LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-2" name="pharmacy" aria-labelledby="pharmacy-2--label" type="radio" value="YGM1E"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-2--label" for="pharmacy-2">YGM1E - MBBM HEALTHCARE TECHNOLOGIES LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-3" name="pharmacy" aria-labelledby="pharmacy-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-3--label" for="pharmacy-3">Other</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class="sc-bcPKBy hhHLxP"><button class="nhsuk-button" aria-disabled="false" type="submit">Release</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;
